### PR TITLE
8320807: [PPC64][ZGC] C1 generates wrong code for atomics

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
Debugging test failures on PPC64 in java/lang/Thread/virtual/stress/Skynet.java#ZGenerational has shown that the ldarx+stdcx_ loop for uncompressed Oops in `LIR_Assembler::atomic_op` is wrong: `__ mr(Rtmp, Robj);` is inside of the ldarx+stdcx_ loop, but must be outside of it. Repeated execution leads to wrong store value.
In addition, zBarrierSetC1.cpp expects `cas_obj` and `xchg` to contain all necessary memory barriers. That doesn't fit to the current PPC64 design which inserts memory barriers on LIR level instead. I've changed this and moved them into the assembler code for all GCs.
While debugging, I have optimized out an unnecessary branch in `ZBarrierSetAssembler::store_barrier_medium`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320807](https://bugs.openjdk.org/browse/JDK-8320807): [PPC64][ZGC] C1 generates wrong code for atomics (**Bug** - P3)

### Reviewers
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**) ⚠️ Review applies to [54a0d25d](https://git.openjdk.org/jdk/pull/16835/files/54a0d25db9fe5b1f371b758ec4fbbc24e099a1ec)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to [54a0d25d](https://git.openjdk.org/jdk/pull/16835/files/54a0d25db9fe5b1f371b758ec4fbbc24e099a1ec)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16835/head:pull/16835` \
`$ git checkout pull/16835`

Update a local copy of the PR: \
`$ git checkout pull/16835` \
`$ git pull https://git.openjdk.org/jdk.git pull/16835/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16835`

View PR using the GUI difftool: \
`$ git pr show -t 16835`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16835.diff">https://git.openjdk.org/jdk/pull/16835.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16835#issuecomment-1828689236)

### CODE REVIEW

File: src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp [hunk 1]
```diff
@@ -2600,6 +2600,13 @@ void LIR_Assembler::emit_compare_and_swap(LIR_OpCompareAndSwap* op) {
     Unimplemented();
   }
 
+  // Volatile load may be followed by Unsafe CAS.
```
Comment (reinrich) [inline src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp:2603]:
```suggestion
  // There might be a volatile load before this Unsafe CAS.
```

File: src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp [hunk 2]
```diff
@@ -2961,9 +2968,24 @@ void LIR_Assembler::atomic_op(LIR_Code code, LIR_Opr src, LIR_Opr data, LIR_Opr
...
+    } else {
+      Robj = data->as_register();
+      if (Robj == dest->as_register()) { // May happen with ZGC.
+        __ mr(Rtmp, Robj);
+        Robj = Rtmp;
+      }
+    }
+  }
+
+  // Volatile load may be followed by Unsafe OP.
```
Comment (reinrich) [inline src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp:2984]:
```suggestion
  // There might be a volatile load before this Unsafe OP.
```

**Review by reinrich — APPROVED** [review]
Looks good. Just two suggestions to improve comments.
Cheers, Richard.

**Comment by TheRealMDoerr** [comment]
Thanks for the reviews! I've improved the comments as suggested.

**Pull request closed without merging**